plugins/meta/firewall/nftables.go

```diff
@@ -0,0 +1,350 @@
+			rule = []string{
+				"add", "rule", nb.targetTable, nb.targetChain,
+				"oifname", fmt.Sprintf("\"%s\"", nb.targetInterface),
+				"ip", "daddr", nb.targetNetwork.String(),
+				"ct", "state", "established,related",
+				"counter", "packets", "0", "bytes", "0", "accept",
+			}
+		} else {
+			rule = []string{
+				"insert", "rule", nb.targetTable, nb.targetChain,
+				"position", fmt.Sprintf("%d", nb.targetHandle),
+				"oifname", fmt.Sprintf("\"%s\"", nb.targetInterface),
+				"ip", "daddr", nb.targetNetwork.String(),
+				"ct", "state", "established,related",
+				"counter", "packets", "0", "bytes", "0", "accept",
+			}
+		}
+		nb.newRules = append(nb.newRules, rule)
+	}
+
+	if !isOutboundInterfaceRuleExists {
+		requireChange = true
+		if len(nb.rules) == 0 {
+			rule = []string{
+				"add", "rule", nb.targetTable, nb.targetChain,
+				"iifname", fmt.Sprintf("\"%s\"", nb.targetInterface),
+				"ip", "saddr", nb.targetNetwork.String(),
+				"counter", "packets", "0", "bytes", "0", "accept",
+			}
+		} else {
+			rule = []string{
+				"insert", "rule", nb.targetTable, nb.targetChain,
+				"position", fmt.Sprintf("%d", nb.targetHandle),
+				"iifname", fmt.Sprintf("\"%s\"", nb.targetInterface),
+				"ip", "saddr", nb.targetNetwork.String(),
+				"counter", "packets", "0", "bytes", "0", "accept",
+			}
+		}
+		nb.newRules = append(nb.newRules, rule)
+	}
+
+	if requireChange {
+		return true
+	}
+	return false
+}
+
+func (nb *nftBackend) addRules() error {
+	if len(nb.newRules) == 0 {
+		return nil
+	}
+	for _, cmdArgs := range nb.newRules {
+		stdoutLines, stderrLines, err := nb.execCommand(cmdArgs)
+		if err != nil {
+			return fmt.Errorf(
+				"encountered error %s: %s (stdout: %s, stderr: %s)",
+				cmdArgs, err, stdoutLines, stderrLines,
+			)
+		}
+	}
+	return nil
+}
+
+func (nb *nftBackend) delRules() error {
+	if len(nb.markedDelRules) == 0 {
+		return nil
+	}
+	return nil
+}
+
+func (nb *nftBackend) isValidInput(result *current.Result) error {
+	if len(result.Interfaces) == 0 {
+		return fmt.Errorf("the data passed to firewall plugin did not contain network interfaces")
+	}
+
+	if result.Interfaces[0].Name == "" {
+		return fmt.Errorf("the data passed to firewall plugin has no bridge name, e.g. cnibr0")
+	}
+
+	nb.targetInterface = result.Interfaces[0].Name
+
+	if len(result.IPs) == 0 {
+		return fmt.Errorf("the data passed to firewall plugin has no IP addresses")
+	}
+
+	if len(result.IPs) != 1 {
+		return fmt.Errorf("the data passed to firewall plugin has more than one IP address")
+	}
+
+	addr := result.IPs[0].Address
+
+	if addr.String() == "" {
+		return fmt.Errorf("the data passed to firewall plugin has empty IP address")
+	}
+
+	if addr.IP.To4() == nil {
+		return fmt.Errorf("the data passed to firewall plugin has non-IPv4 address")
+	}
+
+	_, netAddr, err := net.ParseCIDR(addr.String())
+	if err != nil {
+		return fmt.Errorf("the data passed to firewall plugin has invalid IPv4 address")
+	}
+
+	nb.targetNetwork = netAddr
```

Comment:
targetNetwork here should be an IP, not a CIDR

Reply:
> targetNetwork here should be an IP, not a CIDR

@mars1024 , in this case all you are doing is NAT overloading, why do you need to create rules for each individual IP? It is less intensive to do a calculation for 1 CIDR as opposed to 100s IP addresses.